### PR TITLE
[#159] Add new pool metric pgagroal_query_count and `Query rate` panel

### PR DESCRIPTION
**Pull request by An-DJ**
# Issue
#159 

# Task
* Add `pgagroal_query_count` which characterizes the number of queries.
* Add new panel `Query rate` in Grafana dashboard,

### CODE REVIEW

**Comment by jesperpedersen** [comment]
This would count all client requests as a query - it would be better to just count the actual SQL queries.

See https://www.postgresql.org/docs/devel/protocol-message-formats.html for the message protocol - you will need support for the simple query and the prepared query format. You can use https://github.com/jesperpedersen/pgprtdbg for protocol debugging.

You can likely reuse code from `pipeline_transaction.c` to track message boundaries, but we need guard against this extra overhead -  maybe a `statistics` configuration variable with a default of `on`.

Only `pipeline_session.c` and `pipeline_transaction.c` needs to support this -- `pipeline_perf.c` won't support extended statistics.

**Comment by An-DJ** [comment]
> This would count all client requests as a query - it would be better to just count the actual SQL queries.
> 
> See https://www.postgresql.org/docs/devel/protocol-message-formats.html for the message protocol - you will need support for the simple query and the prepared query format. You can use https://github.com/jesperpedersen/pgprtdbg for protocol debugging.
> 
> You can likely reuse code from `pipeline_transaction.c` to track message boundaries, but we need guard against this extra overhead - maybe a `statistics` configuration variable with a default of `on`.
> 
> Only `pipeline_session.c` and `pipeline_transaction.c` needs to support this -- `pipeline_perf.c` won't support extended statistics.

T

**Comment by An-DJ** [comment]
@jesperpedersen I have check the message kind in `Simple Query` and `Extended Query(prepare-statement)`. 
* For the SimpleQuery, a query can be created by a `Q` message.
* For the PrepareQuery, a prepare query can be created by a `P` message which is used to parse the unsettled sql. Then a execute query can be created by a `E` message which is used to execute the settled sql.

So the query_count should be increased by 'Q' 'P' and 'E'. The newer commit includes that implementation.

File: src/libpgagroal/pipeline_perf.c [hunk 1]
```diff
@@ -97,6 +97,7 @@ performance_client(struct ev_loop *loop, struct ev_io *watcher, int revents)
    wi = (struct worker_io*)watcher;
 
    status = pgagroal_read_socket_message(wi->client_fd, &msg);
+
```
Comment (jesperpedersen) [inline src/libpgagroal/pipeline_perf.c:100]: White space

File: src/libpgagroal/pipeline_session.c [hunk 2]
```diff
@@ -220,6 +220,11 @@ session_client(struct ev_loop *loop, struct ev_io *watcher, int revents)
    {
       if (likely(msg->kind != 'X'))
       {
+         if (msg->kind == 'Q' || msg->kind == 'P' || msg->kind == 'E')
```
Comment (jesperpedersen) [inline src/libpgagroal/pipeline_session.c:223]: `P` is Parse which we shouldn't count as a query, as `B` (Bind) won't need it after 5 statements.

Lets start with `Q` and `E` in this category.

File: src/libpgagroal/pipeline_transaction.c [hunk 3]
```diff
@@ -236,6 +236,11 @@ transaction_client(struct ev_loop* loop, struct ev_io* watcher, int revents)
    {
       if (likely(msg->kind != 'X'))
       {
+         if (msg->kind == 'Q' || msg->kind == 'P' || msg->kind == 'E')
```
Comment (jesperpedersen) [inline src/libpgagroal/pipeline_transaction.c:239]: Same here.

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by jesperpedersen** [comment]
You are missing the reset part for the new metric

**Comment by An-DJ** [comment]
> You are missing the reset part for the new metric

Done. I found that a 'P' message(parse) can be sent by `prepare` statement, and 'B','E'... messages can be sent by  `execute` statement. Maybe the parsing phase should not be considered as query. I have removed it just now.

**Comment by jesperpedersen** [comment]
Thanks for your contribution !

**Pull request merged**